### PR TITLE
Update the PTransform and associated APIs to be less class-based.

### PR DESCRIPTION
**Pull request by robertwb**
This gets us closer to stabilizing the API.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @damccorm

**Review by damccorm — COMMENTED** [review]
This mostly looks good to me - had a few questions/comments + the build is currently failing

File: sdks/typescript/src/apache_beam/coders/coders.ts [hunk 1]
```diff
@@ -38,16 +38,34 @@ interface Class<T> {
...
     }
-    return constructor;
+    if (payload && payload.length > 0) {
+      return constructor(payload, ...components);
+    } else {
+      return constructor(...components);
+    }
+  }
+
+  register(urn: string, constructorOrClass: Class<Coder<any>>) {
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/typescript/src/apache_beam/coders/coders.ts:59]: The intent was to accept both here, but I wasn't able to figure out how to cleanly distinguish between the two. Dropping a TODO for now.

File: sdks/typescript/src/apache_beam/transforms/flatten.ts [hunk 2]
```diff
@@ -22,23 +22,27 @@ import { PCollection } from "../pvalue";
...
+    // TODO: UnionCoder if they're not the same?
+    const coders = new Set(
+      inputs.map((pc) => pipeline.context.getPCollectionCoderId(pc))
+    );
+    const coder =
+      coders.size == 1 ? [...coders][0] : new GeneralObjectCoder<T>();
+    return pipeline.createPCollectionInternal<T>(coder);
   }
+
+  return withName("flatten", expandInternal);
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/typescript/src/apache_beam/transforms/flatten.ts:45]: Done.

**Comment by robertwb** [comment]
Thanks. Addressed your comments and fixed the build. PTAL.

File: sdks/typescript/src/apache_beam/transforms/pardo.ts [hunk 3]
```diff
@@ -140,96 +135,77 @@ export class ParDo<InputT, OutputT, ContextT = undefined> extends PTransform<
...
+      tags.push(options.unknownTagName!);
+    }
+    if (options.knownTags === undefined) {
+      options.knownTags = tags;
+    }
 
-// TODO: (Typescript) Is it possible to type that this takes
-// PCollection<{a: T, b: U, ...}> to {a: PCollection<T>, b: PCollection<U>, ...}
-// Seems to requires a cast inside expandInternal. But at least the cast is contained there.
-export class Split2<T extends { [key: string]: unknown }> extends PTransform<
```
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/typescript/src/apache_beam/transforms/pardo.ts:197]: Cool, SGTM

**Review by damccorm — APPROVED** [review]
Thanks, LGTM

**Pull request merged**